File: Python/383_ransom-note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        magazine_map = {}
        for elm in magazine:
            if elm not in magazine_map:
                magazine_map[elm] = 1
            else:
                magazine_map[elm] +=1
        
        for elm in ransomNote:
            if elm not in magazine_map or magazine_map[elm] <= 0:
                return False
            else:
                magazine_map[elm] -= 1
        return True
        

File: Java/383_ransom-note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        Map<Character, Integer> magazineMap = new HashMap<>();
        for(char c : magazine.toCharArray()){
            if (magazineMap.containsKey(c)){
                magazineMap.put(c, magazineMap.get(c)+1);
            }else{
                magazineMap.put(c, 1);
            }
        }
        for(char c : ransomNote.toCharArray()){
            if(!magazineMap.containsKey(c) ||magazineMap.get(c) == 0 ){
                return false;
            }else{
                magazineMap.put(c, magazineMap.get(c)-1);
            }
        }
        return true;
    }
}